web/client-api/src/main/java/io/deephaven/web/client/api/tree/JsTreeTable.java

```diff
@@ -602,6 +374,213 @@ private TicketAndPromise<?> makeView(TicketAndPromise<?> prevTicket) {
+                return new TreeRowImpl(subscription, index);
+            }
+
+            @JsProperty
+            public double getTreeSize() {
+                return treeSize;
+            }
+        }
+
+        public class TreeRowImpl extends SubscriptionRow implements TreeViewportData.TreeRow {
+
+            public TreeRowImpl(WebBarrageSubscription subscription, long index) {
+                super(subscription, rowStyleColumn, index);
+            }
+
+            @Override
+            public boolean isExpanded() {
+                return barrageSubscription.getData(index, rowExpandedCol.getIndex()).uncheckedCast() == Boolean.TRUE;
+            }
+
+            @Override
+            public boolean hasChildren() {
+                return barrageSubscription.getData(index, rowExpandedCol.getIndex()).uncheckedCast() != null;
+            }
+
+            @Override
+            public int depth() {
+                return Js.coerceToInt(barrageSubscription.getData(index, rowDepthCol.getIndex()));
+            }
+
+            public void appendKeyData(Object[][] keyTableData, double action) {
+                int i;
+                for (i = 0; i < keyColumns.length; i++) {
+                    Js.<JsArray<Any>>cast(keyTableData[i]).push(keyColumns.getAt(i).get(this));
+                }
+                Js.<JsArray<Double>>cast(keyTableData[i++]).push((double) depth());
+                Js.<JsArray<Double>>cast(keyTableData[i++]).push(action);
+            }
+
+            @Override
+            public Any get(Column column) {
+                Column sourceColumn = sourceColumns.get(column.getName());
+                if (sourceColumn == null) {
+                    // no constituent column, call super
+                    return super.get(column);
+                }
+                if (barrageSubscription.getData(index, rowDepthCol.getIndex()).asInt() != constituentDepth()) {
+                    // not at constituent depth, call super
+                    return super.get(column);
+                }
+                // read source col instead
+                return super.get(sourceColumn);
+            }
+
+            @Override
+            public Format getFormat(Column column) {
+                Column sourceColumn = sourceColumns.get(column.getName());
+                if (sourceColumn == null) {
+                    // no constituent column, call super
+                    return super.getFormat(column);
+                }
+                if (barrageSubscription.getData(index, rowDepthCol.getIndex()).asInt() != constituentDepth()) {
+                    // not at constituent depth, call super
+                    return super.getFormat(column);
+                }
+                // read source col instead
+                return super.getFormat(sourceColumn);
+            }
+        }
+
+        // TODO move to superclass and check on viewport change
+        private RangeSet serverViewport;
```

Comment:
TODO ticket follow up?

Reply:
Comment removed, this serves a new purpose since the comment was first added.